### ISSUE TITLE
update_item does not work correctly with SET update expressions

### ISSUE THREAD

**Issue by ndamclean** [spulec/moto#3336]
**moto version:**
\>=1.3.15

**Issue**
The `update_item` function does not work as expected. You should be able to use `SET` in an "update expression" to add an attribute to an item if the attribute does not already exist. This was working correctly in `moto` version `1.3.14`.

> An expression that defines one or more attributes to be updated, the action to be performed on them, and new values for them.
> 
> The following action values are available for UpdateExpression.
> 
>     SET - Adds one or more attributes and values to an item.
>           If any of these attributes already exist, they are replaced by the new values.
>           You can also use SET to add or subtract from an attribute that is of type Number.
>           For example: SET myNum = myNum + :val 
https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateItem.html#API_UpdateItem_RequestSyntax

**Code to reproduce error:**
```
import boto3
import moto
import os
from mock import patch


@moto.mock_dynamodb2
def test_update_item():
    with patch.dict(os.environ, {
        'AWS_DEFAULT_REGION': 'us-west-2',
        'AWS_ACCESS_KEY_ID': 'AKIA0000000000000000',
        'AWS_SECRET_ACCESS_KEY': '0000000000000000000000000000000000000000,'
    }):
        dynamodb = boto3.resource('dynamodb')

        dynamodb.create_table(
            TableName='test',
            AttributeDefinitions=[
                {
                    'AttributeName': 'key',
                    'AttributeType': 'S'
                },
            ],
            KeySchema=[
                {
                    'AttributeName': 'key',
                    'KeyType': 'HASH'
                }
            ],
            BillingMode='PAY_PER_REQUEST',
        )

        table = dynamodb.Table('test')

        table.put_item(Item={'key': 'key1'})

        table.update_item(
            Key={'key': 'key1'},
            UpdateExpression='SET other_attribute = other_value',
        )
```

**Error traceback:**
```
Traceback (most recent call last):
  File "tmp.py", line 43, in <module>
    test_update_item()
  File "/home/nmclean/.virtualenvs/tmp/lib/python3.6/site-packages/moto/core/models.py", line 100, in wrapper
    result = func(*args, **kwargs)
  File "tmp.py", line 39, in test_update_item
    UpdateExpression='SET other_attribute = other_value',
  File "/home/nmclean/.virtualenvs/tmp/lib/python3.6/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/home/nmclean/.virtualenvs/tmp/lib/python3.6/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
  File "/home/nmclean/.virtualenvs/tmp/lib/python3.6/site-packages/botocore/client.py", line 337, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/nmclean/.virtualenvs/tmp/lib/python3.6/site-packages/botocore/client.py", line 656, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ValidationException) when calling the UpdateItem operation: The provided expression refers to an attribute that does not exist in the item
```

**Comment by bblommers** [comment]
Thanks for raising this @ndamclean - that's a very common scenario, not sure why moto doesn't cover that.

 Marking this as a bug.

**Comment by bblommers** [comment]
Hi @ndamclean, just did some debugging - it looks like AWS behaves the same way.
It only works (both in AWS and moto) when specifying the ExpressionAttributeNames/Values:
```
res = table.update_item(
        Key={'key': 'key1'},
        UpdateExpression='SET #att=:val',
        ExpressionAttributeNames={"#att": "other_attribute"},
        ExpressionAttributeValues={":val": "other_value"}
    )
```

**Comment by bblommers** [comment]
Going to close this, as I think it's working as expected. Let me know if I've missed anything.

**Comment by n27jain** [comment]
> Hi @ndamclean, just did some debugging - it looks like AWS behaves the same way. It only works (both in AWS and moto) when specifying the ExpressionAttributeNames/Values:
> 
> ```
> res = table.update_item(
>         Key={'key': 'key1'},
>         UpdateExpression='SET #att=:val',
>         ExpressionAttributeNames={"#att": "other_attribute"},
>         ExpressionAttributeValues={":val": "other_value"}
>     )
> ```

How do we do this for multiple attributes?
I need to do this for 3 attributes.

**Comment by bblommers** [comment]
If the question is how to set multiple attributes at the same time, @n27jain, this page may be useful:
https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html

TLDR:
You can simply list multiple expressions, separated by comma, like this:
```
UpdateExpression='SET #att1=:val, #att2=:val2',
ExpressionAttributeNames={"#att1": "..", "#att2": ".."},
ExpressionAttributeValues={":val1": "...", ":val1": "..."}
```
